scipy/fft/tests/test_helper.py

```diff
@@ -391,55 +391,63 @@ def test_uneven_dims(self, xp):
 class TestFFTFreq:
 
     def test_definition(self, xp):
-        device = SCIPY_DEVICE
-        try:
-            x = xp.asarray([0, 1, 2, 3, 4, -4, -3, -2, -1],
-                           dtype=xp.float64, device=device)
-            x2 = xp.asarray([0, 1, 2, 3, 4, -5, -4, -3, -2, -1],
-                            dtype=xp.float64, device=device)
-        except TypeError:
-            x = xp.asarray([0, 1, 2, 3, 4, -4, -3, -2, -1], dtype=xp.float64)
-            x2 = xp.asarray([0, 1, 2, 3, 4, -5, -4, -3, -2, -1],
-                            dtype=xp.float64)
-
-        y = xp.asarray(9 * fft.fftfreq(9, xp=xp), dtype=xp.float64)
+        xp_test = array_namespace(xp.empty(0))
+
+        x = xp.asarray([0, 1, 2, 3, 4, -4, -3, -2, -1], dtype=xp.float64)
+        x2 = xp.asarray([0, 1, 2, 3, 4, -5, -4, -3, -2, -1], dtype=xp.float64)
+
+        y = 9 * fft.fftfreq(9, xp=xp_test)
+        assert array_namespace(y) is xp_test
+        y = xp_test.astype(y, xp.float64)
         xp_assert_close(y, x)
-        y = xp.asarray(9 * xp.pi * fft.fftfreq(9, xp.pi, xp=xp), dtype=xp.float64)
+
+        y = xp_test.astype(9 * xp.pi * fft.fftfreq(9, xp.pi, xp=xp_test), xp.float64)
         xp_assert_close(y, x)
 
-        y = xp.asarray(10 * fft.fftfreq(10, xp=xp), dtype=xp.float64)
+        y = xp_test.astype(10 * fft.fftfreq(10, xp=xp_test), xp.float64)
         xp_assert_close(y, x2)
-        y = xp.asarray(10 * xp.pi * fft.fftfreq(10, xp.pi, xp=xp), dtype=xp.float64)
+
+        y = xp_test.astype(10 * xp.pi * fft.fftfreq(10, xp.pi, xp=xp_test), xp.float64)
         xp_assert_close(y, x2)
 
+    def test_device(self, xp):
+        xp_test = array_namespace(xp.empty(0))
+        devices = get_xp_devices(xp)
+        for d in devices:
+            y = fft.fftfreq(9, xp=xp_test, device=d)
+            x = xp_test.empty(0, device=d)
+            assert device(y) == device(x)
+
 
-@skip_xp_backends('array_api_strict', 'cupy',
-                   reasons=['fft not yet implemented by array-api-strict',
-                            'cupy.fft not yet implemented by array-api-compat'])
 class TestRFFTFreq:
 
     def test_definition(self, xp):
-        device = SCIPY_DEVICE
-        try:
-            x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64, device=device)
-            x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64, device=device)
-        except TypeError:
-            # work around the `device` keyword not being implemented in numpy yet
-            x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64)
-            x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64)
-
-        y = xp.asarray(9 * fft.rfftfreq(9, xp=xp), dtype=xp.float64)
+        xp_test = array_namespace(xp.empty(0))
+
+        x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64)
+        x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64)
+
+        y = 9 * fft.rfftfreq(9, xp=xp_test)
+        assert array_namespace(y) is xp_test
```

Comment:
What's going on with the new `xp_test` thing above and below and this separate call instead of using the namespace checks we built in to `xp_assert_..` functions? We had a way to disable the dtype checks as well I thought, so I guess I'm trying to wrap my head around the extra activity/coercion/checks.

If CuPy `device` handling is the issue should we just skip it for now instead of this dance?

```diff
 class TestRFFTFreq:
 
+    @skip_xp_backends("cupy",
+                      reasons=["no device arg support"])
     def test_definition(self, xp):
-        xp_test = array_namespace(xp.empty(0))
-
         x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64)
         x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64)
 
-        y = 9 * fft.rfftfreq(9, xp=xp_test)
-        assert array_namespace(y) is xp_test
-        y = xp_test.astype(y, xp.float64)
-        xp_assert_close(y, x)
+        y = 9 * fft.rfftfreq(9, xp=xp)
+        xp_assert_close(y, x, check_dtype=False, check_namespace=True)
 
-        y = xp_test.astype(9 * xp.pi * fft.rfftfreq(9, xp.pi, xp=xp_test), xp.float64)
-        xp_assert_close(y, x)
+        y = 9 * xp.pi * fft.rfftfreq(9, xp.pi, xp=xp)
+        xp_assert_close(y, x, check_dtype=False)
 
-        y = xp_test.astype(10 * fft.rfftfreq(10, xp=xp_test), xp.float64)
-        xp_assert_close(y, x2)
+        y = 10 * fft.rfftfreq(10, xp=xp)
+        xp_assert_close(y, x2, check_dtype=False)
 
-        y = xp_test.astype(10 * xp.pi * fft.rfftfreq(10, xp.pi, xp=xp_test), xp.float64)
-        xp_assert_close(y, x2)
+        y = 10 * xp.pi * fft.rfftfreq(10, xp.pi, xp=xp)
+        xp_assert_close(y, x2, check_dtype=False)
```

Reply:
> What's going on with the new xp_test thing above and below and this separate call instead of using the namespace checks we built in to xp_assert_.. functions? We had a way to disable the dtype checks as well I thought, so I guess I'm trying to wrap my head around the extra activity/coercion/checks.

I suppose we can just turn off the dtype check yeah. I was trying to check both the namespace and the dtype at once, but maybe we just introduce a separate `test_dtypes`.

I don't think we need to skip this for CuPy - we may need to skip the new `test_device` tests, if we decide to include CuPy support in the new helper.

Reply:
Ok, in local testing, CuPy was failing because of `device` pass-through I think.

Reply:
Interesting. I'm away from my GPU, so feel free to post the errors. Or I can check later.

Reply:
For the modified version of the test above, I see the failure below with the skip removed. My assumption was that all the extra shims in the test were guarding against the CuPy `device` stuff, so I think that's what I'm getting confused about re: performing testing gymnastics vs. actually just providing a support shim in the source proper if we want to test that functionality.

```
scipy/fft/tests/test_helper.py:431: in test_definition
    y = 9 * fft.rfftfreq(9, xp=xp)
        self       = <scipy.fft.tests.test_helper.TestRFFTFreq object at 0x7fc490f214d0>
        x          = array([0., 1., 2., 3., 4.])
        x2         = array([0., 1., 2., 3., 4., 5.])
        xp         = <module 'cupy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'>
scipy/fft/_helper.py:216: in rfftfreq
    return xp.fft.rfftfreq(n, d=d, device=device)
E   TypeError: rfftfreq() got an unexpected keyword argument 'device'
        d          = 1.0
        device     = None
        n          = 9
        xp         = <module 'cupy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'>
===================================================================================== short test summary info =====================================================================================
FAILED scipy/fft/tests/test_helper.py::TestRFFTFreq::test_definition[cupy] - TypeError: rfftfreq() got an unexpected keyword argument 'device'

```

Reply:
That's the part I think is confusing/debatable, and I'm not sure I agree just yet. If CuPy isn't compatible with the API here, why don't we skip until it is?

If other folks actually prefer the approach you've used here, I think comment(s) are needed to explain, because these test shims are messy to understand.

Reply:
Ah, I suppose that is option (3) then. I think I was tunnel visioned on getting the test to pass (to at least check that our implementation is correct), but now that I'm confident that what we have is correct, I think you're right that we should just skip.

Reply:
I've reworked based on your comments, let me know what you think